File: backend/routes/automation.py
"""
Terminal automation and tar creation routes.
Uses xdotool on Ubuntu to type text into the active terminal window.
"""
from fastapi import APIRouter, HTTPException
import subprocess
import shutil
import os
import time

router = APIRouter(prefix="/api/automation", tags=["automation"])


def _xdotool_available() -> bool:
    return shutil.which("xdotool") is not None


def _run(cmd: list[str], timeout: int = 10) -> dict:
    try:
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout,
        )
        return {
            "ok": result.returncode == 0,
            "stdout": result.stdout.strip(),
            "stderr": result.stderr.strip(),
            "returncode": result.returncode,
        }
    except subprocess.TimeoutExpired:
        return {"ok": False, "stdout": "", "stderr": "Command timed out", "returncode": -1}
    except Exception as e:
        return {"ok": False, "stdout": "", "stderr": str(e), "returncode": -1}


@router.get("/status")
async def automation_status():
    """Check whether xdotool is available on this system."""
    available = _xdotool_available()
    return {
        "xdotool_available": available,
        "platform": os.name,
        "note": "xdotool only works on Linux/Ubuntu with X11. Install with: sudo apt install xdotool",
    }


@router.post("/type")
async def type_text(data: dict):
    """
    Type text into the currently focused window using xdotool.
    Optionally append a newline (press_enter=true).
    """
    text = data.get("text", "")
    press_enter = data.get("press_enter", False)
    delay_ms = int(data.get("delay_ms", 0))

    if not text and not press_enter:
        raise HTTPException(400, "text or press_enter required")

    if not _xdotool_available():
        raise HTTPException(503, "xdotool not installed. Run: sudo apt install xdotool")

    if delay_ms > 0:
        time.sleep(delay_ms / 1000)

    results = []

    if text:
        r = _run(["xdotool", "type", "--clearmodifiers", "--delay", "20", "--", text])
        results.append(r)
        if not r["ok"]:
            raise HTTPException(500, f"xdotool type failed: {r['stderr']}")

    if press_enter:
        time.sleep(0.05)
        r = _run(["xdotool", "key", "Return"])
        results.append(r)
        if not r["ok"]:
            raise HTTPException(500, f"xdotool key Return failed: {r['stderr']}")

    return {"ok": True, "steps": results}


@router.post("/key")
async def press_key(data: dict):
    """Press a key combination using xdotool (e.g., Return, ctrl+c)."""
    key = data.get("key", "")
    if not key:
        raise HTTPException(400, "key required")

    if not _xdotool_available():
        raise HTTPException(503, "xdotool not installed. Run: sudo apt install xdotool")

    r = _run(["xdotool", "key", key])
    if not r["ok"]:
        raise HTTPException(500, f"xdotool key failed: {r['stderr']}")

    return {"ok": True}


@router.post("/tar")
async def create_tar(data: dict):
    """
    Create a tar archive of a directory.
    body: { "source_path": "/path/to/repo", "output_path": "/path/to/output.tar" }
    """
    source = data.get("source_path", "").strip()
    output = data.get("output_path", "").strip()

    if not source or not output:
        raise HTTPException(400, "source_path and output_path required")

    if not os.path.exists(source):
        raise HTTPException(404, f"Source path not found: {source}")

    source = os.path.abspath(source)
    output = os.path.abspath(output)

    # Ensure output directory exists
    os.makedirs(os.path.dirname(output) or ".", exist_ok=True)

    r = _run(["tar", "cf", output, "-C", os.path.dirname(source), os.path.basename(source)], timeout=120)

    if not r["ok"]:
        raise HTTPException(500, f"tar failed: {r['stderr']}")

    size = os.path.getsize(output) if os.path.exists(output) else 0
    return {
        "ok": True,
        "output_path": output,
        "size_bytes": size,
        "size_mb": round(size / (1024 * 1024), 2),
    }

File: backend/routes/automation.js
const { Router }    = require('express');
const { execFileSync } = require('child_process');
const fs   = require('fs');
const path = require('path');

const router = Router();
const sleep = ms => new Promise(r => setTimeout(r, ms));

function xdotoolAvailable() {
  if (process.platform === 'win32') return false;
  try { execFileSync('which', ['xdotool'], { stdio: 'pipe' }); return true; }
  catch { return false; }
}

function run(cmd, args, timeout = 10000) {
  try {
    const stdout = execFileSync(cmd, args, { timeout, encoding: 'utf-8', stdio: 'pipe' });
    return { ok: true, stdout: stdout.trim(), stderr: '', returncode: 0 };
  } catch (e) {
    return { ok: false, stdout: '', stderr: (e.stderr || e.message || '').toString().trim(), returncode: e.status || -1 };
  }
}

router.get('/status', (_req, res) => {
  res.json({
    xdotool_available: xdotoolAvailable(),
    platform: process.platform,
    note: 'xdotool only works on Linux/Ubuntu with X11. Install with: sudo apt install xdotool',
  });
});

router.post('/type', async (req, res) => {
  const { text = '', press_enter = false, delay_ms = 0 } = req.body;
  if (!text && !press_enter) return res.status(400).json({ detail: 'text or press_enter required' });
  if (!xdotoolAvailable()) return res.status(503).json({ detail: 'xdotool not installed. Run: sudo apt install xdotool' });

  if (delay_ms > 0) await sleep(delay_ms);

  const steps = [];
  if (text) {
    const r = run('xdotool', ['type', '--clearmodifiers', '--delay', '20', '--', text]);
    steps.push(r);
    if (!r.ok) return res.status(500).json({ detail: `xdotool type failed: ${r.stderr}` });
  }
  if (press_enter) {
    await sleep(50);
    const r = run('xdotool', ['key', 'Return']);
    steps.push(r);
    if (!r.ok) return res.status(500).json({ detail: `xdotool key Return failed: ${r.stderr}` });
  }
  res.json({ ok: true, steps });
});

router.post('/key', (req, res) => {
  const { key = '' } = req.body;
  if (!key) return res.status(400).json({ detail: 'key required' });
  if (!xdotoolAvailable()) return res.status(503).json({ detail: 'xdotool not installed. Run: sudo apt install xdotool' });
  const r = run('xdotool', ['key', key]);
  if (!r.ok) return res.status(500).json({ detail: `xdotool key failed: ${r.stderr}` });
  res.json({ ok: true });
});

router.post('/tar', (req, res) => {
  let { source_path = '', output_path = '' } = req.body;
  source_path = source_path.trim();
  output_path = output_path.trim();
  if (!source_path || !output_path)
    return res.status(400).json({ detail: 'source_path and output_path required' });
  if (!fs.existsSync(source_path))
    return res.status(404).json({ detail: `Source path not found: ${source_path}` });

  source_path = path.resolve(source_path);
  output_path = path.resolve(output_path);
  fs.mkdirSync(path.dirname(output_path), { recursive: true });

  const r = run('tar', ['cf', output_path, '-C', path.dirname(source_path), path.basename(source_path)], 120000);
  if (!r.ok) return res.status(500).json({ detail: `tar failed: ${r.stderr}` });

  const size = fs.existsSync(output_path) ? fs.statSync(output_path).size : 0;
  res.json({ ok: true, output_path, size_bytes: size, size_mb: +(size / 1024 / 1024).toFixed(2) });
});

module.exports = router;
